File: Data Structures/5. Swift/k. Heaps.swift
struct Heap<T> {
    private var elements: [T]
    private var isMinHeap: Bool
    private var comparator: (T, T) -> Bool

    init(isMinHeap: Bool = false, comparator: @escaping (T, T) -> Bool) {
        self.elements = []
        self.isMinHeap = isMinHeap
        self.comparator = comparator
    }

    // MARK: - Insertion

    /// Insert an element into the heap.
    mutating func insert(_ element: T) {
        elements.append(element)
        heapifyUp()
    }

    private mutating func heapifyUp() {
        var currentIndex = elements.count - 1
        while currentIndex > 0 {
            let parentIndex = (currentIndex - 1) / 2

            if shouldSwap(parentIndex, currentIndex) {
                elements.swapAt(parentIndex, currentIndex)
                currentIndex = parentIndex
            } else {
                break
            }
        }
    }

    // MARK: - Removal

    /// Remove the root element from the heap.
    mutating func remove() -> T? {
        guard !elements.isEmpty else { return nil }

        if elements.count == 1 {
            return elements.removeLast()
        }

        let root = elements[0]
        elements[0] = elements.removeLast()
        heapifyDown()

        return root
    }

    private mutating func heapifyDown() {
        var currentIndex = 0

        while hasLeftChild(currentIndex) {
            var smallestChildIndex = leftChildIndex(currentIndex)

            if hasRightChild(currentIndex) && shouldSwap(rightChildIndex(currentIndex), smallestChildIndex) {
                smallestChildIndex = rightChildIndex(currentIndex)
            }

            if shouldSwap(smallestChildIndex, currentIndex) {
                elements.swapAt(smallestChildIndex, currentIndex)
                currentIndex = smallestChildIndex
            } else {
                break
            }
        }
    }

    // MARK: - Peek

    /// Peek at the root element without removing it.
    func peek() -> T? {
        return elements.first
    }

    // MARK: - Helper Methods

    private func leftChildIndex(_ parentIndex: Int) -> Int {
        return 2 * parentIndex + 1
    }

    private func rightChildIndex(_ parentIndex: Int) -> Int {
        return 2 * parentIndex + 2
    }

    private func hasLeftChild(_ index: Int) -> Bool {
        return leftChildIndex(index) < elements.count
    }

    private func hasRightChild(_ index: Int) -> Bool {
        return rightChildIndex(index) < elements.count
    }

    private func shouldSwap(_ firstIndex: Int, _ secondIndex: Int) -> Bool {
        if isMinHeap {
            return comparator(elements[secondIndex], elements[firstIndex])
        } else {
            return comparator(elements[firstIndex], elements[secondIndex])
        }
    }

    // MARK: - Build Heap

    /// Builds a heap from an array of elements.
    mutating func buildHeap(_ elements: [T]) {
        self.elements = elements
        for i in stride(from: (elements.count / 2) - 1, through: 0, by: -1) {
            heapifyDown(i)
        }
    }

    // MARK: - Replace Root

    /// Replace the root element with a new element.
    mutating func replaceRoot(with element: T) {
        guard !elements.isEmpty else {
            insert(element)
            return
        }

        elements[0] = element
        heapifyDown()
    }

    // MARK: - Remove at Index

    /// Remove an element at a specific index from the heap.
    mutating func remove(at index: Int) -> T? {
        guard index < elements.count else { return nil }

        if index == elements.count - 1 {
            return elements.removeLast()
        }

        elements.swapAt(index, elements.count - 1)
        let removedElement = elements.removeLast()
        heapifyDown(index)

        return removedElement
    }

    // MARK: - Sort

    /// Sort the heap in-place (ascending order).
    mutating func sort() {
        for i in stride(from: elements.count - 1, through: 1, by: -1) {
            elements.swapAt(0, i)
            heapifyDown(0, endIndex: i)
        }
    }

    // MARK: - Index for Element

    /// Get the index of a specific element, if it exists in the heap.
    func index(of element: T) -> Int? {
        return index(of: element, startingAt: 0)
    }

    private func index(of element: T, startingAt index: Int) -> Int? {
        guard index < elements.count else { return nil }

        if elements[index] == element {
            return index
        }

        if let leftChildIndex = leftChildIndex(index), let result = index(of: element, startingAt: leftChildIndex) {
            return result
        }

        if let rightChildIndex = rightChildIndex(index), let result = index(of: element, startingAt: rightChildIndex) {
            return result
        }

        return nil
    }

    // ... Add more heap operations as needed ...
}

File: data_structures/1_cpp/k_heaps.cpp
#include <iostream>
#include <vector>
#include <functional>

template <typename T>
class Heap {
private:
    std::vector<T> elements;
    bool isMinHeap;
    std::function<bool(T, T)> comparator;

public:
    Heap(bool isMinHeap = false, std::function<bool(T, T)> comparator = [](T a, T b) { return a < b; }) {
        this->isMinHeap = isMinHeap;
        this->comparator = comparator;
    }

    // MARK: - Insertion

    void insert(T element) {
        elements.push_back(element);
        heapifyUp();
    }

    void heapifyUp() {
        int currentIndex = elements.size() - 1;
        while (currentIndex > 0) {
            int parentIndex = (currentIndex - 1) / 2;

            if (shouldSwap(parentIndex, currentIndex)) {
                std::swap(elements[parentIndex], elements[currentIndex]);
                currentIndex = parentIndex;
            } else {
                break;
            }
        }
    }

    // MARK: - Removal

    T remove() {
        if (elements.empty()) {
            throw std::runtime_error("Heap is empty");
        }

        T root = elements[0];
        elements[0] = elements.back();
        elements.pop_back();
        heapifyDown();

        return root;
    }

    void heapifyDown() {
        int currentIndex = 0;

        while (hasLeftChild(currentIndex)) {
            int smallestChildIndex = leftChildIndex(currentIndex);

            if (hasRightChild(currentIndex) && shouldSwap(rightChildIndex(currentIndex), smallestChildIndex)) {
                smallestChildIndex = rightChildIndex(currentIndex);
            }

            if (shouldSwap(smallestChildIndex, currentIndex)) {
                std::swap(elements[smallestChildIndex], elements[currentIndex]);
                currentIndex = smallestChildIndex;
            } else {
                break;
            }
        }
    }

    // MARK: - Peek

    T peek() const {
        if (elements.empty()) {
            throw std::runtime_error("Heap is empty");
        }

        return elements[0];
    }

    // MARK: - Helper Methods

    int leftChildIndex(int parentIndex) const {
        return 2 * parentIndex + 1;
    }

    int rightChildIndex(int parentIndex) const {
        return 2 * parentIndex + 2;
    }

    bool hasLeftChild(int index) const {
        return leftChildIndex(index) < elements.size();
    }

    bool hasRightChild(int index) const {
        return rightChildIndex(index) < elements.size();
    }

    bool shouldSwap(int firstIndex, int secondIndex) const {
        if (isMinHeap) {
            return comparator(elements[secondIndex], elements[firstIndex]);
        } else {
            return comparator(elements[firstIndex], elements[secondIndex]);
        }
    }

    // MARK: - Build Heap

    void buildHeap(const std::vector<T>& elements) {
        this->elements = elements;
        for (int i = (elements.size() / 2) - 1; i >= 0; --i) {
            heapifyDown(i);
        }
    }

    // MARK: - Replace Root

    void replaceRoot(T element) {
        if (elements.empty()) {
            insert(element);
            return;
        }

        elements[0] = element;
        heapifyDown();
    }

    // MARK: - Remove at Index

    T removeAtIndex(int index) {
        if (index >= elements.size()) {
            throw std::out_of_range("Index out of range");
        }

        if (index == elements.size() - 1) {
            T removed = elements.back();
            elements.pop_back();
            return removed;
        }

        std::swap(elements[index], elements.back());
        T removed = elements.back();
        elements.pop_back();
        heapifyDown(index);

        return removed;
    }

    // MARK: - Sort

    void sort() {
        for (int i = elements.size() - 1; i >= 1; --i) {
            std::swap(elements[0], elements[i]);
            heapifyDown(0, i);
        }
    }

    // MARK: - Index for Element

    int index(const T& element) const {
        return index(element, 0);
    }

    int index(const T& element, int startIndex) const {
        if (startIndex >= elements.size()) {
            return -1;
        }

        if (elements[startIndex] == element) {
            return startIndex;
        }

        int leftIndex = leftChildIndex(startIndex);
        int rightIndex = rightChildIndex(startIndex);

        if (leftIndex < elements.size()) {
            int result = index(element, leftIndex);
            if (result != -1) {
                return result;
            }
        }

        if (rightIndex < elements.size()) {
            return index(element, rightIndex);
        }

        return -1;
    }

    // ... Add more heap operations as needed ...
};
